SUR-232-Pagination-Component

HimanshuSangale opened the pull request:
### Description

Added Pagination component.

### Screenshots

https://d.pr/v/DfNWzg

### Checklist:

-   [x] My code is tested
-   [ ] My code passes the PHPCS tests
-   [ ] I've created the npm build.
-   [ ] My code follows accessibility standards <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
-   [x] My code has proper inline documentation <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
-   [ ] I've included any necessary tests <!-- if applicable -->
-   [ ] I've included developer documentation <!-- if applicable -->
-   [ ] I've added proper labels to this pull request <!-- if applicable -->

ravindra114 commented on src/components/pagination/pagination.jsx:
```diff
@@ -78,73 +89,71 @@ const PaginationLink = ({
 
-const PaginationPrevious = ({ onClick = () => { }, className, ...props }) => {
-	const { size, disabled } = usePageContext();
+const PaginationPrevious = ( { onClick = () => {}, className, ...props } ) => {
```
No need to set onClick in the body again

ravindra114 commented on src/components/pagination/pagination.jsx:
```diff
@@ -78,73 +89,71 @@ const PaginationLink = ({
 
-const PaginationNext = ({ onClick = () => { }, className, ...props }) => {
-	const { size, disabled } = usePageContext();
+const PaginationNext = ( { onClick = () => {}, className, ...props } ) => {
```
No need to set onClick in the body again

ravindra114 reviewed (commented):
You can merge small components
`
const PaginationButton = ({ icon, ariaLabel, onClick, className, ...props }) => (
  <PaginationLink
    aria-label={ariaLabel}
    icon={icon}
    onClick={onClick}
    className={className}
    {...props}
  />
);
`

then use it like below
`
const PaginationPrevious = (props) => (
  <PaginationButton icon={<ChevronLeft />} ariaLabel="Go to previous page" {...props} />
);

const PaginationNext = (props) => (
  <PaginationButton icon={<ChevronRight />} ariaLabel="Go to next page" {...props} />
);
`

vrundakansara commented on src/components/pagination/readme.md:
````diff
@@ -0,0 +1,59 @@
+### Basic Example
+
+```jsx
+import { Label } from '@bsf/force-ui';
+
+const App = () => (
+    <div>
+        <Label size='xs'>Vrunda <MapPinPlusInside/> Kansara </Label>
+        <Label size='sm'>Label</Label>
+        <Label size='sm' variant='disabled'><MapPinPlusInside/> Label</Label>
+        <Label size='md' variant="help">Another Label <MapPinPlusInside/></Label>
+        <Label variant="error">An error label with a <a href="#">link</a>.</Label>
+    </div>
+);
+
+export default App;
````
Add valid example

vrundakansara commented on src/components/pagination/readme.md:
```diff
@@ -0,0 +1,59 @@
+# Label
```
Modify the entire file to match pagination

vrundakansara commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,165 @@
+		<PaginationLink
+			aria-label="Go to next page"
+			className={ cn(
+				// sizeClassNames[size].icon,
+				// disabled && disabledClassNames.icon,
```
Are these classes not required? If so we can get rid of commented code.

vrundakansara commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,165 @@
+	<ul
+		ref={ ref }
+		className={ cn(
+			'm-0 w-full flex justify-center flex-row items-center gap-1',
```
We might need classes that disable the list style as well, this will not work properly in some pages, where the stylings are added at body level.

HimanshuSangale replied:
Fixed.

HimanshuSangale commented:
I've added the readme file and removed the commented code. @vrundakansara

jaieds reviewed (changes requested):
I've found some design mismatches and other issues. Please watch the video and the screenshot.

👉 Video: https://share.cleanshot.com/lhZpty1D
👉 Screenshot: https://d.pr/i/HIyL0M

jaieds commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,142 @@
+		/>
+	);
+} );
+PaginationItem.displayName = 'PaginationItem';
```
Follow other components for the Sub-components display name.

HimanshuSangale commented:
@ravindra114 
- Refined the code and reduced number for components for easier usage ✅ 

@jaieds 
- Fixed sizing issue by using aspect ratio instead of separate paddings for all. ✅ 
- Font size and color are already as per figma design and handled by Button component. ✅

jaieds reviewed (changes requested):
@HimanshuSangale
- Active, focus state colors are not accurate.
- Icon size is not accurate.
- Arrow button size is smaller than other buttons.
- The 3 dots button should not be clickable.

Watch this video for clarity: https://d.pr/v/08BVuU

HimanshuSangale commented:
@jaieds 
- Fixed focus state backgroudn color, Active state colors are as per Figma and also discussed with Akbar regarding same. ✅ 
- Icon size fixed. Now button component handles the icon size and looks appropriate. ✅ https://d.pr/i/pYNikE
- 3 dots (Ellipses) are now only hoverable and focusable, not clickable, only cursor visible on hover. ✅ https://d.pr/v/1OtkKq

ravindra114 commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,165 @@
+		>
+			<PaginationButton
+				icon={ <ChevronRight /> }
+				ariaLabel="Go to next page"
```
ariaLabel -> aria-label

ravindra114 commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,165 @@
+		>
+			<PaginationButton
+				icon={ <ChevronLeft /> }
+				ariaLabel="Go to previous page"
```
ariaLabel -> aria-label

JelenaTakac reviewed (changes requested):
Please check the video, does not change active state properly.
https://bsf.d.pr/v/WRJZAs

HimanshuSangale commented:
@ravindra114 
- Fixed above issue ✅ 

@JelenaTakac 
- Intentionally removed the click action so the focus effect is properly visible. ✅

ravindra114 reviewed (changes requested):
https://d.pr/i/ZiESL4
the structure seems incorrect, ... should not be at end

HimanshuSangale commented:
@ravindra114 
- The order of page buttons changed in the storybook.

JelenaTakac reviewed (approved):
Tested.
https://bsf.d.pr/v/q5EdgM

BabuTechBsf reviewed (changes requested):
Pagination
UI
Ellisis should have no effect or onclicks or hover.
I don’t see xs variation in figma documentation? 

Story:
Custom width code visible: Width of component should be taken from inner elements. why 400px used? if necessary you can use decorators.
 Exampls are basic not required. Have some better examples which can’t be altered easyly  with args etc..
 all args are not provided in doc, like isactive, clasnames etc..
 
 Can we give an option for navigation like Icon/text/button
multiple onClick={() => {}} present in code.
https://d.pr/i/yIi2eg - looks like the implementation is wrong here. achieve the square shape with padings, if text is passed also the items should look good.

console log statements ??
 <Pagination.Previous onClick={() => console.log('Previous page')} />
      <Pagination.Item isActive onClick={() => console.log('Go to page 1')}> 1 </Pagination.Item>
      <Pagination.Item onClick={() => console.log('Go to page 2')}> 2 </Pagination.Item>
      <Pagination.Ellipsis onClick={() => console.log('Show more pages')} />
      <Pagination.Item onClick={() => console.log('Go to page 10')}> 10 </Pagination.Item>
    <Pagination.Next onClick={() => console.log('Next page')} />

HimanshuSangale commented:
@BabuTechBsf 
- All effects removed from ellipses. ✅ 
- xs variation is given in Atoms documentation https://d.pr/i/y6C0jQ, as mentioned in previous PR change suggestions.
- Added examples which can be manipulated by args easily, added all args. ✅ 
- Option to pass icon or text for Previous and Next buttons provided.

BabuTechBsf commented:
> Pagination UI Ellisis should have no effect or onclicks or hover. I don’t see xs variation in figma documentation?
> 
> Story: Custom width code visible: Width of component should be taken from inner elements. why 400px used? if necessary you can use decorators. Exampls are basic not required. Have some better examples which can’t be altered easyly with args etc.. all args are not provided in doc, like isactive, clasnames etc.. even for subcomponents
> 
> Can we give an option for navigation like Icon/text/button multiple onClick={() => {}} present in code. https://d.pr/i/yIi2eg - looks like the implementation is wrong here. achieve the square shape with padings, if text is passed also the items should look good.
> 
> console log statements ?? <Pagination.Previous onClick={() => console.log('Previous page')} /> <Pagination.Item isActive onClick={() => console.log('Go to page 1')}> 1 </Pagination.Item> <Pagination.Item onClick={() => console.log('Go to page 2')}> 2 </Pagination.Item> <Pagination.Ellipsis onClick={() => console.log('Show more pages')} /> <Pagination.Item onClick={() => console.log('Go to page 10')}> 10 </Pagination.Item> <Pagination.Next onClick={() => console.log('Next page')} />

@HimanshuSangale  Pls add your response all the points mentioned

HimanshuSangale commented:
> @BabuTechBsf
> 
> * All effects removed from ellipses. ✅
> * xs variation is given in Atoms documentation https://d.pr/i/y6C0jQ, as mentioned in previous PR change suggestions. ✅
> * Added examples which can be manipulated by args easily, added all args. ✅
> * Option to pass icon or text for Previous and Next buttons provided. ✅
> * Fixed button shape using padding. ✅
> * Removed all unnecessary code. ✅

@BabuTechBsf 
- Removed custom width from storybook. ✅ 
- Removed all console log code ✅ 
- Removed onclick completely from the implementation. ✅

jaieds commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,174 @@
+			href={ href }
+			className="text-none no-underline bg-transparent p-0 m-0 border-none"
+		>
+			<Button
```
Why are we using a button inside the anchor tag?

HimanshuSangale replied:
Fixed ✅

jaieds commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,174 @@
+	const { size, disabled } = usePageContext();
+
+	return (
+		<a
+			href={ href }
+			className="text-none no-underline bg-transparent p-0 m-0 border-none"
+		>
```
If I want to set a `target` or `rel` how do I do that?

HimanshuSangale replied:
Added ✅

jaieds commented on src/components/pagination/pagination.stories.js:
```diff
@@ -0,0 +1,163 @@
+				<Pagination.Ellipsis />
+				<Pagination.Item
+					href={ args.href }
+					isActive={ false || args.isActive }
```
Set `target="_self"` and `href="#"` to all pagination items (only in story).

HimanshuSangale replied:
Done ✅

jaieds reviewed (changes requested):
The ellipsis dimension is different from others.

HimanshuSangale commented:
@jaieds 
- Done all above changes. ✅ 
- Fixed ellipses padding. ✅

jaieds commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,199 @@
+
+	return (
+		<Button
+			tag="a"
```
Instead of setting it statically, we can leave this to the user, and we can set the tag value to a by default if not passed by users. Also, parameters related to the `a` tag should not be set by default. If a developer wants to use React Router Link as the tag value, in that case it will create an issue.

jaieds commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,197 @@
+} ) => {
+	const { size, disabled } = usePageContext();
+
+	const validTag =
```
If we are allowed to pass only string then it won't be possible to pass any other component. I want to use the Link (from React Router) component as the Tag will I be able to do that?

HimanshuSangale replied:
I'll add a check.

jaieds commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,197 @@
+					isActive={ isActive }
+					disabled={ disabled }
+					className={ className }
+					href={ href }
```
```suggestion
```

jaieds commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,197 @@
+PaginationContent.displayName = 'Pagination.Content';
+
+const PaginationItem = forwardRef(
+	( { isActive = false, href = '', className, children, ...props }, ref ) => {
```
```suggestion
	( { isActive = false, className, children, ...props }, ref ) => {
```

jaieds commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,197 @@
+				icon={ icon }
+				aria-label="Go to next page"
+				className={ sizeClassNames[ size ].icon }
+				href={ href }
+				target={ target }
+				rel={ rel }
```
```suggestion
```

jaieds commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,197 @@
+
+const PaginationNext = ( {
+	icon = <ChevronRight />,
+	href = '',
+	target = '_self',
+	rel,
```
```suggestion
```

jaieds commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,197 @@
+				icon={ icon }
+				aria-label="Go to previous page"
+				className={ sizeClassNames[ size ].icon }
+				href={ href }
+				target={ target }
+				rel={ rel }
```
```suggestion
```

jaieds commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,197 @@
+
+const PaginationPrevious = ( {
+	icon = <ChevronLeft />,
+	href = '',
+	target = '_self',
+	rel,
```
```suggestion
```

HimanshuSangale commented:
@jaieds 
- Now you can pass component to tag, like Link. Added checks for function and string both and it defaults to 'a' tag when none available.

jaieds commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,181 @@
+
+	// Check if `tag` is a valid HTML tag (string) or a React component (function), default to 'a' if not
+	const validTag =
+		tag && ( typeof tag === 'string' || typeof tag === 'function' )
```
The type of a component is `object`. So this logic will not work. Instead, we can set the `tag` default value to `a`. And if the developer passes anything (it can be a string to a component) it will use that component.

HimanshuSangale replied:
I'll make the changes.

HimanshuSangale replied:
@jaieds 
- Made the desired changes. ✅

jaieds commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,175 @@
+	);
+};
+
+const PaginationPrevious = ( { icon = <ChevronLeft />, props } ) => {
```
```suggestion
const PaginationPrevious = ( { icon = <ChevronLeft />, ...props } ) => {
```

jaieds commented on src/components/pagination/pagination.jsx:
```diff
@@ -0,0 +1,175 @@
+};
+PaginationPrevious.displayName = 'Pagination.Previous';
+
+const PaginationNext = ( { icon = <ChevronRight />, props } ) => {
```
```suggestion
const PaginationNext = ( { icon = <ChevronRight />, ...props } ) => {
```

jaieds reviewed (changes requested):
The `href` and `target` props are not forwarded to the button component. Please check.

HimanshuSangale commented:
@jaieds 
- Made some changes to storybook code. ✅

jaieds reviewed (changes requested):
Causing a layout shift issue because of the active item border. Instead of a border, we can use an outline and ring the class name.

The ring and outline don't cause any change in dimension.

Check the video (outline & ring class name use): https://d.pr/v/JpQxfP

Pagination component layout shift issue: https://d.pr/v/JgHdEO

HimanshuSangale commented:
@jaieds 
- Implemented outline ✅
